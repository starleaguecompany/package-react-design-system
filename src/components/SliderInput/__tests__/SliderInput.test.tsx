import * as React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SliderInput } from '../../../index';
import { Menu } from '@starleaguecompany/react-icons';

describe('SliderInput', () => {
  test('should render correctly with no props', () => {
    render(<SliderInput />);

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByTestId('SliderInput')).toHaveClass('container');
  });

  test('should render correctly with attributes', () => {
    render(<SliderInput id="test-id" className="test-class" />);

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByTestId('SliderInput')).toHaveClass('test-class');
    expect(screen.getByRole('textbox')).toHaveAttribute('id', 'test-id');
  });

  test('should render correctly with value prop', () => {
    const { container } = render(<SliderInput value={10} />);
    const handler = container.querySelector('.handler');

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByRole('textbox')).toHaveValue('10');
    expect(screen.getByText('10')).toBeVisible;
    expect(handler).toHaveStyle('margin-left: 10%');
  });

  test('should render correctly with defaultValue prop', () => {
    const { container } = render(<SliderInput defaultValue={10} />);
    const handler = container.querySelector('.handler');

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByRole('textbox')).toHaveValue('10');
    expect(screen.getByText('10')).toBeVisible;
    expect(handler).toHaveStyle('margin-left: 10%');
  });

  test('should render correctly with postfix prop', () => {
    render(<SliderInput value={10} postfix="₽" />);

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByText('₽')).toBeVisible();
    expect(screen.getByText('₽')).toHaveClass('inputPostfix');
    expect(screen.getByTestId('SliderInput')).toHaveTextContent('10 ₽');
  });

  test('should render correctly with disabled prop', () => {
    const { container } = render(<SliderInput disabled />);
    const disabledElements = container.querySelectorAll('.disabled');

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByTestId('TextInput')).toHaveClass('disabled');
    expect(screen.getByRole('textbox')).toBeDisabled();
    expect(screen.getByRole('textbox')).toHaveClass('disabled');
    expect(disabledElements.length).toEqual(4);
  });

  test('should render correctly with invalid prop', () => {
    const { container } = render(<SliderInput invalid />);
    const invalidElements = container.querySelectorAll('.invalid');

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByTestId('TextInput')).toHaveClass('invalid');
    expect(screen.getByRole('textbox')).toHaveClass('invalid');
    expect(invalidElements.length).toEqual(2);
  });

  test('should render correctly with readOnly prop', () => {
    const { container } = render(<SliderInput value={50} readOnly />);
    const inner = container.querySelector('.inner');

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByRole('textbox')).toHaveAttribute('readonly');
    userEvent.click(inner);
    expect(screen.getByRole('textbox')).toHaveValue('50');
  });

  test('should render correctly with max prop', () => {
    const { container, rerender } = render(<SliderInput max={100} value={150} />);
    const handler = container.querySelector('.handler');

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(handler).toHaveStyle('margin-left: 100%');
    rerender(<SliderInput max={100} value={100} />);
    expect(handler).toHaveStyle('margin-left: 100%');
    rerender(<SliderInput max={100} value={99} />);
    expect(handler).toHaveStyle('margin-left: 99%');
  });

  test('should render correctly with min prop', () => {
    const { container, rerender } = render(<SliderInput min={100} value={50} max={200} />);
    const handler = container.querySelector('.handler');

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(handler).toHaveStyle('margin-left: 0%');
    rerender(<SliderInput min={100} value={100} max={200} />);
    expect(handler).toHaveStyle('margin-left: 0%');
    rerender(<SliderInput min={100} value={101} max={200} />);
    expect(handler).toHaveStyle('margin-left: 1%');
  });

  test('should execute onChange callback on input value', async () => {
    const onChangeCallback = jest.fn();
    const { container } = render(<SliderInput value={5} onChange={onChangeCallback} />);
    const handler = container.querySelector('.handler');

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(onChangeCallback).toBeCalledTimes(0);
    userEvent.click(screen.getByRole('textbox'));
    userEvent.keyboard('0');
    await waitFor(() => expect(onChangeCallback).toBeCalledTimes(1));
    expect(screen.getByRole('textbox')).toHaveValue('50');
    expect(handler).toHaveStyle('margin-left: 50%');
  });

  test('should render correctly with loading prop', () => {
    const { rerender } = render(<SliderInput loading />);

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByTestId('SliderInput')).toContainElement(screen.getByTestId('Spinner'));
    rerender(<SliderInput />);
    expect(screen.getByTestId('SliderInput')).not.toContainElement(screen.queryByTestId('Spinner'));
  });

  test('should render correctly with autoFocus prop', () => {
    render(<SliderInput autoFocus />);

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByTestId('TextInput')).toHaveClass('focused');
    expect(screen.getByRole('textbox')).toHaveFocus();
  });

  test('should render correctly with icon prop', () => {
    render(<SliderInput icon={<Menu />} />);

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    expect(screen.getByTestId('SliderInput')).toContainElement(screen.getByTestId('Icon'));
    expect(screen.getByTestId('Icon')).toContainElement(screen.getByTestId('Menu'));
  });

  test('should render correctly on update value', () => {
    const onChangeCallback = jest.fn();
    const { rerender } = render(<SliderInput value={50} onChange={onChangeCallback} />);

    expect(screen.getByTestId('SliderInput')).toMatchSnapshot();
    rerender(<SliderInput value={90} onChange={onChangeCallback} />);
    expect(screen.getByText('90')).toBeVisible();
    expect(screen.getByRole('textbox')).toHaveValue('90');
    expect(onChangeCallback).toBeCalledTimes(0);
  });
});
