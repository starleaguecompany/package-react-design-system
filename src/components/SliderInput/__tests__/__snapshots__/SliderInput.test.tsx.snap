// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SliderInput should execute onChange callback on input value 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container"
      data-qa="TextInput"
    >
      <div
        class="inputSegment centered"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          >
            5
          </div>
          <input
            autocomplete="off"
            class="input filled"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value="5"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 5%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly on update value 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container"
      data-qa="TextInput"
    >
      <div
        class="inputSegment centered"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          >
            50
          </div>
          <input
            autocomplete="off"
            class="input filled"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value="50"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 50%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with attributes 1`] = `
<div
  class="test-class container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input"
            id="test-id"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with autoFocus prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container focused active"
      data-qa="TextInput"
    >
      <div
        class="inputSegment centered"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input active"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with defaultValue prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container"
      data-qa="TextInput"
    >
      <div
        class="inputSegment centered"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          >
            10
          </div>
          <input
            autocomplete="off"
            class="input filled"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value="10"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 10%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with disabled prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container disabled"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input disabled"
            disabled=""
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner disabled"
    >
      <div
        class="track disabled"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with icon prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container with-icon"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
      <span
        class="icon container size-20 shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="Menu"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 5h12M4 10h12M4 15h12"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with invalid prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container invalid"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input invalid"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with loading prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container with-icon"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
      <span
        class="icon container size-20"
        data-qa="Spinner"
      >
        <svg
          class="svg"
          viewBox="0 0 150 150"
        >
          <circle
            class="circle circle-1"
            cx="75"
            cy="75"
            r="60"
          />
          <circle
            class="circle circle-2"
            cx="75"
            cy="75"
            r="60"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with max prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container"
      data-qa="TextInput"
    >
      <div
        class="inputSegment centered"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          >
            100
          </div>
          <input
            autocomplete="off"
            class="input filled"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value="100"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 100%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with min prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container"
      data-qa="TextInput"
    >
      <div
        class="inputSegment centered"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          >
            100
          </div>
          <input
            autocomplete="off"
            class="input filled"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value="100"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with no props 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with postfix prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container"
      data-qa="TextInput"
    >
      <div
        class="inputSegment centered"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          >
            10
          </div>
          <input
            autocomplete="off"
            class="input filled"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value="10"
          />
          <div
            class="inputPostfix"
          >
             
            ₽
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 10%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with readOnly prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container"
      data-qa="TextInput"
    >
      <div
        class="inputSegment centered"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          >
            50
          </div>
          <input
            autocomplete="off"
            class="input filled"
            inputmode="decimal"
            readonly=""
            spellcheck="false"
            type="text"
            value="50"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 50%;"
      />
    </div>
  </div>
</div>
`;

exports[`SliderInput should render correctly with value prop 1`] = `
<div
  class="container"
  data-qa="SliderInput"
>
  <div
    class="container"
    data-qa="NumberInput"
  >
    <div
      class="container"
      data-qa="TextInput"
    >
      <div
        class="inputSegment centered"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          >
            10
          </div>
          <input
            autocomplete="off"
            class="input filled"
            inputmode="decimal"
            spellcheck="false"
            type="text"
            value="10"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="slider"
    data-qa="Slider"
  >
    <div
      class="inner"
    >
      <div
        class="track"
      />
      <span
        class="handler"
        style="margin-left: 10%;"
      />
    </div>
  </div>
</div>
`;
